### PR TITLE
support multiple packages per changefile

### PR DESCRIPTION
**Pull request by jcreamer898**
In large repos when you're changing a bunch of packages at once, it can be kind of expensive to dump a bunch of JSON files into a single folder.

It's easy enough to just group the change files into single files expressing multiple packages per file.

Essentially, there would be a new `multiPackageChanges` option that would generate change files like...

```json
{
"changes": [{
  "packageName": "beachball",
  "type": "minor",
  "dependentChangeType": "patch",
  "email": "jcreamer@microsoft.com",
  "comment": "support multiple changes per changefile"
  }]
}
```

And there's a bit of confusion on my end as to why each `ChangeInfo` needs its own `comment` as I think that `comment` field is generated from the `yarn change` command? So perhaps we can bubble that `comment` field up into the `summary` as seen above.

Other than supporting multiple packages, in theory nothing else here functionally changes, other than I added the `human-id` library to give the change files a shorter and cleaner name.

### CODE REVIEW

**Comment by VincentBailly** [comment]
Do the suggestion of using toml have to be coupled with the suggestion to use a single files?
Coupling the two may dilute the core proposal, which is to use only one change file per PR.

**Comment by jcreamer898** [comment]
Good call, maybe I can accomplish the same thing, but with an array of JSON in one file.

**Comment by jcreamer898** [comment]
Updated @VincentBailly

**Review by ecraig12345 — CHANGES_REQUESTED** [review]
Thanks for the contribution! I have some questions though. Also I'd definitely like to see some tests covering the new feature.

> In large repos when you're changing a bunch of packages at once, it can be kind of expensive to dump a bunch of JSON files into a single folder.

Expensive in what sense? If you mean it makes beachball commands slow to run, I'm not sure that the number of change files is actually the root cause (there are a couple other possibilities I'm aware of).

> And there's a bit of confusion on my end as to why each ChangeInfo needs its own comment as I think that comment field is generated from the yarn change command? So perhaps we can bubble that comment field up into the summary as seen above.

It depends on the scenario--sometimes using the same message across multiple packages makes sense, but other times it's good to be more granular. By default beachball asks for an individual message for each package, and the `--message` option is available as an override to give a single message for all packages.

> Other than supporting multiple packages, in theory nothing else here functionally changes, other than I added the human-id library to give the change files a shorter and cleaner name.

Whether this is actually an improvement is debatable, so I'd prefer it be discussed/handled separately.

**Comment by jcreamer898** [comment]
Thanks for the reply @ecraig12345!

I added a few e2e tests to the PR now.

Expensive in this case was for git. In 1JS in MS Office, we have a folder with lots and lots of change files (that's a different topic), in any case, I did some research that shows that having a lot of files in one folder is really bad for git history and repo size, perf, etc. https://gist.github.com/jcreamer898/54f91122a5b64907eb02da0dc3c78b26

Thanks for the clarity on the message thing. It didn't end up mattering for my change.

Ken and I discussed the `human-id` change, and he was on board with it.

Thanks again!

**Review by kenotron — DISMISSED** [review]

**Comment by jcreamer898** [comment]
Need new approval here @kenotron

File: src/changefile/writeChangeFiles.ts
Comment (ecraig12345) [inline src/changefile/writeChangeFiles.ts:6]: Would it be possible to add a repo option `humanId?: boolean` and only use `human-id` if that's enabled? I don't think it's desirable to name change files with random words in all cases (I wouldn't want this), and regardless, suddenly changing to that might catch people off guard.

File: src/types/ChangeInfo.ts
Comment (ecraig12345) [inline src/types/ChangeInfo.ts:23]: nit, can you use an interface here to match the convention in this repo?

**Review by VincentBailly — APPROVED** [review]

**Review by ecraig12345 — CHANGES_REQUESTED** [review]
Looked at this again and I have some more comments.

Also does this cause any changes to the *default* way change files are written? I'm not seeing this option enabled for beachball itself (unless I'm missing something) but it appears to be used in this PR's change file.

File: src/types/BeachballOptions.ts
Comment (ecraig12345) [inline src/types/BeachballOptions.ts:45]: This name is a bit verbose, especially if it has to be specified on the command line. I wonder if it would be possible to come up with something more concise but still clear? 

Whatever the name is, it should probably be added to `RepoOptions` in addition or instead (IMO this would make more sense to enable in a config file than on the command line), and documented in `docs/overview/configuration.md`.

Comment (kenotron) [reply src/types/BeachballOptions.ts:45]: I agree with the comment here about putting in "RepoOptions" and generally be checked into a beachball.config.js rather than in a CLI...

File: src/changefile/writeChangeFiles.ts
Comment (jcreamer898) [reply src/changefile/writeChangeFiles.ts:6]: I got this idea from https://github.com/atlassian/changesets/blob/eae72367e9cb8e708a1d441770dbc70930d4e1ad/packages/write/src/index.ts#L17 and personally find it much more readable than a guid, but that's up to you and @kenotron.

File: src/types/BeachballOptions.ts
Comment (jcreamer898) [reply src/types/BeachballOptions.ts:45]: Sounds great. Thanks.

File: src/changefile/readChangeFiles.ts
Comment (jcreamer898) [reply src/changefile/readChangeFiles.ts:75]: Oo, nice.

File: src/types/ChangeInfo.ts
Comment (jcreamer898) [reply src/types/ChangeInfo.ts:23]: Good call, not sure what I was thinking. Ha.

File: src/changefile/writeChangeFiles.ts
Comment (kenotron) [reply src/changefile/writeChangeFiles.ts:6]: I think this is completely subjective. The ONLY reason why I wouldn't do it is the addition of yet another dependency. Maybe we should just revert it for now and add it as a neat nice-to-have feature?

Comment (jcreamer898) [reply src/changefile/writeChangeFiles.ts:6]: I'll create a separate PR to talk about that. Don't wanna block this one on this.

**Comment by jcreamer898** [comment]
Updated the option to `groupChanges`, and added docs, updated tests.

File: src/changefile/writeChangeFiles.ts
Comment (ecraig12345) [reply src/changefile/writeChangeFiles.ts]: > I think this is completely subjective. The ONLY reason why I wouldn't do it is the addition of yet another dependency. Maybe we should just revert it for now and add it as a neat nice-to-have feature?

Definitely agree that it's subjective (and I'm not concerned about the extra dep). It just seems like an especially polarizing type of subjective, as demonstrated by both of you liking it and me thinking "eew this is weird."

File: change/change-chatty-cloths-develop.json
Comment (ecraig12345) [inline change/change-chatty-cloths-develop.json:2]: One more question (sorry)--does this PR change the default way change files are written? I didn't see the new option being enabled for beachball itself, but it looks like this change file is using the new format.

Comment (jcreamer898) [reply change/change-chatty-cloths-develop.json:2]: The update supports the multi-change format in the readChangefile, but the default in writeChangefile will still be regular way it is now.

**Comment by jcreamer898** [comment]
Also, don't apologize, I truly appreciate your thoroughness. :)

**Review by ecraig12345 — APPROVED** [review]

**Comment by ecraig12345** [comment]
Thanks for all the work on this! I think I'm actually going to enable it in our repo as well because it will make PR reviews easier.

**Comment by jcreamer898** [comment]
Nice! How do I actually merge the change now? I have a not authorized message. :)

**Comment by ecraig12345** [comment]
Oh sorry, didn't realize you didn't have permission

**Pull request merged**
